File: typescript/packages/core/src/sandbox/remote/session.ts
import { SandboxId, SnapshotId } from "../../log";
import { err, ok } from "../../result";
import { admitExec } from "../admit";
import { isRefusal } from "../context";
import type {
  ExecOutput,
  Failure,
  FileFailure,
  ReleaseFailure,
  SandboxContext,
  SandboxSession,
  SnapshotData,
  Trees,
} from "../protocol";
import { type ReadFailure, treeHash } from "../trees";
import { byteStream, joined } from "./bytes";
import type { SandboxDriver } from "./driver";
import { FenceRefused, guarded, messageOf } from "./fence";
import {
  checkReadScript,
  checkWriteScript,
  EXPORT_TREE_SCRIPT,
  execScript,
  FILE_EXIT,
  importTreeScript,
  sandboxPath,
  stdinPath,
  treePath,
  WORKSPACE,
} from "./scripts";

// One remote sandbox as a SandboxSession (spec/api.json): every operation fences through the
// driver's transport, and everything above the raw provider calls is the shared POSIX kit.

/** A kit script's exit code and output, collected: its outputs are small. */
export async function runScript(
  driver: SandboxDriver,
  id: string,
  script: string,
): Promise<{
  readonly exit: number;
  readonly stdout: Uint8Array;
  readonly stderr: string;
}> {
  const out = byteStream();
  const err = byteStream();
  const started = await driver.run(id, script, {
    stdout: out.push,
    stderr: err.push,
  });
  const stdout = joined(out.chunks);
  const stderr = joined(err.chunks);
  const exit = await started.exit.finally(() => {
    out.end();
    err.end();
  });
  return {
    exit,
    stdout: await stdout,
    stderr: new TextDecoder().decode(await stderr),
  };
}

/**
 * The manifest hash of the sandbox's /workspace through its export, inside a kit operation:
 * undefined when it can't be read, and a refused fence thrown so the operation reports it.
 */
export async function measured(
  session: Pick<Trees, "exportTree">,
  context: SandboxContext,
): Promise<string | undefined> {
  const hash = await treeHash(session, context);
  if (hash.ok) return hash.value;
  if (isRefusal<ReadFailure["code"]>(hash.error))
    throw new FenceRefused(hash.error);
  return undefined;
}

/** A script started with its output streamed, never joined (exec, the tree export). */
async function streamed(
  driver: SandboxDriver,
  id: string,
  script: string,
  processKey?: string,
): Promise<ExecOutput> {
  const stdout = byteStream();
  const stderr = byteStream();
  const started = await driver.run(
    id,
    script,
    { stdout: stdout.push, stderr: stderr.push },
    processKey,
  );
  const exit = started.exit.finally(() => {
    stdout.end();
    stderr.end();
  });
  // Read after both streams end; a transport failure mid-stream rejects it then.
  exit.catch(() => undefined);
  return { exit_code: exit, stdout: stdout.chunks, stderr: stderr.chunks };
}

const unavailable = (error: unknown) =>
  ({ code: "unavailable", message: messageOf(error) }) as const;

const FILE_CODES = new Map<number, FileFailure["code"]>([
  [FILE_EXIT.not_found, "not_found"],
  [FILE_EXIT.is_directory, "is_directory"],
  [FILE_EXIT.permission_denied, "permission_denied"],
]);

/** Runs a file check; a typed failure, or undefined when the path may be used. */
async function checked(
  driver: SandboxDriver,
  id: string,
  script: string,
  path: string,
): Promise<FileFailure | undefined> {
  const { exit } = await runScript(driver, id, script);
  if (exit === 0) return undefined;
  return {
    code: FILE_CODES.get(exit) ?? "unavailable",
    message: `${path}: exit ${exit}`,
  };
}

export function remoteSession(
  driver: SandboxDriver,
  provider: string,
  id: string,
): SandboxSession & Trees {
  const sandboxId = SandboxId.parse(id);
  // A confirmed driver's answer. Otherwise a best-effort kill that never claims it worked: a
  // descendant can drop out of the process the provider tracks, so only an operator can
  // settle it (termination: unconfirmed).
  const terminate: SandboxSession["terminate"] = (processKey, context) =>
    guarded<
      "terminated" | "already_exited" | "unknown",
      Failure<"unavailable">
    >(
      context,
      async () => {
        if (driver.termination === "confirmed")
          return ok(await driver.terminate(id, processKey));
        await driver.stopProcess(id, processKey);
        return ok("unknown");
      },
      unavailable,
    );

  const exec: SandboxSession["exec"] = async (given, context, givenOptions) => {
    // Before `guarded`, which would turn a caller bug into a typed `unavailable`.
    const { command, options } = admitExec(given, givenOptions);
    return guarded<
      ExecOutput,
      Failure<"timeout" | "invalid_path" | "unavailable">
    >(
      context,
      async () => {
        const cwd = sandboxPath(options.cwd ?? WORKSPACE);
        if (cwd === undefined)
          return err({ code: "invalid_path", message: `cwd ${options.cwd}` });
        if (options.stdin !== undefined)
          await driver.write(id, stdinPath(options.processKey), options.stdin);
        const script = execScript({
          command,
          cwd,
          env: options.env,
          processKey: options.processKey,
          stdin: options.stdin !== undefined,
        });
        // The deadline (timeoutMs) is the sandbox layer's: execute() reports it as a timeout
        // and calls terminate (sandbox/exec.ts).
        return ok(await streamed(driver, id, script, options.processKey));
      },
      unavailable,
    );
  };

  const upload: SandboxSession["upload"] = (path, data, context) =>
    guarded<void, FileFailure>(
      context,
      async () => {
        const at = sandboxPath(path);
        if (at === undefined)
          return err({ code: "invalid_path", message: path });
        const refused = await checked(driver, id, checkWriteScript(at), path);
        if (refused !== undefined) return err(refused);
        await driver.write(id, at, data);
        return ok(undefined);
      },
      unavailable,
    );

  const download: SandboxSession["download"] = (path, context) =>
    guarded<Uint8Array, FileFailure>(
      context,
      async () => {
        const at = sandboxPath(path);
        if (at === undefined)
          return err({ code: "invalid_path", message: path });
        const refused = await checked(driver, id, checkReadScript(at), path);
        return refused === undefined
          ? ok(await driver.read(id, at))
          : err(refused);
      },
      unavailable,
    );

  const exportTree: Trees["exportTree"] = (context) =>
    guarded<ExecOutput, Failure<"unavailable">>(
      context,
      async () => ok(await streamed(driver, id, EXPORT_TREE_SCRIPT)),
      unavailable,
    );

  // Uploaded as one file, then extracted by the sandbox's own tar (spec/schema/README.md,
  // "Sandbox image"); the builder already masked every mode to 0o777.
  const importTree: Trees["importTree"] = (tar, context) =>
    guarded<void, Failure<"unavailable">>(
      context,
      async () => {
        const path = treePath();
        await driver.write(id, path, await joined(tar));
        const run = await runScript(driver, id, importTreeScript(path));
        return run.exit === 0
          ? ok(undefined)
          : err({
              code: "unavailable",
              message: `the import into ${WORKSPACE} exited ${run.exit}: ${run.stderr}`,
            });
      },
      unavailable,
    );

  const capture = driver.snapshot;
  // Only a provider-owned whole-sandbox boundary makes a capture quiescent. The tree hashed
  // before and after must also match, or a writer ran around it (not_quiescent).
  const snapshot: SandboxSession["snapshot"] = (operationKey, context) =>
    guarded<SnapshotData, Failure<"not_quiescent" | "unavailable" | "timeout">>(
      context,
      async () => {
        if (capture === undefined || capture.quiescence === "unconfirmed")
          return err({
            code: "unavailable",
            message: `${provider} has no confirmed quiescent snapshot`,
          });
        const before = await measured({ exportTree }, context);
        const made = await capture.take(id, operationKey);
        const after = await measured({ exportTree }, context);
        if (before === undefined || before !== after) {
          await driver.deleteSnapshot(made.ref);
          return err({
            code: before === undefined ? "unavailable" : "not_quiescent",
            message: `the tree of ${id} changed around the capture`,
          });
        }
        // The boundary covered the whole sandbox: every process in it was frozen, or ended.
        const whole = [id];
        return ok({
          snapshot_id: SnapshotId.parse(made.ref),
          provider,
          sandbox_id: sandboxId,
          capture_class: "filesystem",
          expires_at: made.expiresAt,
          manifest_hash: before,
          quiesced:
            capture.quiescence === "paused"
              ? { frozen: whole, stopped: [], excluded: [] }
              : { frozen: [], stopped: whole, excluded: [] },
        });
      },
      unavailable,
    );

  const close: SandboxSession["close"] = (context) =>
    guarded<void, ReleaseFailure>(
      context,
      async () => {
        await driver.kill(id);
        return ok(undefined);
      },
      (error) => ({ code: "release_failed", message: messageOf(error) }),
    );

  return {
    id: sandboxId,
    exec,
    terminate,
    upload,
    download,
    snapshot,
    close,
    exportTree,
    importTree,
  };
}

File: python/src/threads/sandbox/remote/session.py
"""One remote sandbox as a `SandboxSession` (spec/api.json): every operation fences through the
driver's transport, and everything above the raw provider calls is the shared POSIX kit
(adapters/sandboxes/posix.py)."""

from collections.abc import AsyncIterable, Awaitable, Callable, Mapping, Sequence
from typing import assert_never

from threads.adapters.sandboxes import posix
from threads.adapters.sandboxes.fence import dispatch
from threads.log import SnapshotData
from threads.loop.tools import Termination
from threads.result import Err, Ok
from threads.sandbox.protocol import (
    NO_ENV,
    ExecOutput,
    SandboxContext,
    SandboxError,
    SandboxId,
    invalid_path,
    is_refusal,
)
from threads.sandbox.remote.driver import (
    Capture,
    Confirmed,
    FileError,
    SandboxDriver,
    Taken,
    Unconfirmed,
)
from threads.sandbox.trees import tree_hash


async def call[T](
    driver: SandboxDriver, context: SandboxContext, op: Callable[[], Awaitable[T]]
) -> Ok[T] | Err[SandboxError]:
    """One driver operation under `context`: a refused fence is its typed refusal, a FileError
    its named failure, anything else what the driver's `classify` says it is."""

    def classify(error: Exception) -> SandboxError | None:
        return error.error if isinstance(error, FileError) else driver.classify(error)

    return await dispatch(context, op, classify)


def released(error: SandboxError) -> Err[SandboxError]:
    """A failed release: a refusal as it is, anything else release_failed."""
    return Err(error if is_refusal(error) else SandboxError("release_failed", error.message))


class RemoteSession:
    def __init__(self, driver: SandboxDriver, provider: str, ident: str) -> None:
        self._driver, self._provider = driver, provider
        self._id = SandboxId(ident)

    @property
    def id(self) -> SandboxId:
        return self._id

    async def exec(  # noqa: PLR0913 - the options spec/api.json names
        self,
        command: Sequence[str],
        context: SandboxContext,
        *,
        process_key: str,
        cwd: str = "/workspace",
        env: Mapping[str, str] = NO_ENV,
        timeout_ms: int | None = None,
        stdin: bytes | None = None,
    ) -> Ok[ExecOutput] | Err[SandboxError]:
        # The deadline is the sandbox layer's (run_exec): it reports a timeout and terminates.
        return await posix.run(
            self, command, context, process_key=process_key, cwd=cwd, env=env, stdin=stdin
        )

    async def start(
        self,
        argv: Sequence[str],
        env: Mapping[str, str],
        cwd: str,
        context: SandboxContext,
        process_key: str | None = None,
    ) -> Ok[ExecOutput] | Err[SandboxError]:
        """posix.Primitives: the provider's own exec."""
        return await call(
            self._driver, context, lambda: self._driver.run(self._id, argv, env, cwd, process_key)
        )

    async def terminate(
        self, process_key: str, context: SandboxContext
    ) -> Ok[Termination] | Err[SandboxError]:
        """A confirmed driver's answer; otherwise a best-effort kill that never claims it
        worked: a descendant can drop out of what the provider tracks (the effect parks)."""
        match self._driver.termination:
            case Confirmed(terminate=terminate):
                return await call(self._driver, context, lambda: terminate(self._id, process_key))
            case Unconfirmed(stop_process=stop):
                stopped = await call(self._driver, context, lambda: stop(self._id, process_key))
                return stopped if isinstance(stopped, Err) else Ok("unknown")
            case _:
                assert_never(self._driver.termination)

    async def upload(
        self, path: str, data: bytes, context: SandboxContext
    ) -> Ok[None] | Err[SandboxError]:
        bad = invalid_path(path)
        if bad is not None:
            return bad
        return await call(self._driver, context, lambda: self._driver.write(self._id, path, data))

    async def download(self, path: str, context: SandboxContext) -> Ok[bytes] | Err[SandboxError]:
        bad = invalid_path(path)
        if bad is not None:
            return bad
        return await call(self._driver, context, lambda: self._driver.read(self._id, path))

    async def export_tree(self, context: SandboxContext) -> Ok[ExecOutput] | Err[SandboxError]:
        return await posix.export_tree(self, context)

    async def import_tree(
        self, tar: AsyncIterable[bytes], context: SandboxContext
    ) -> Ok[None] | Err[SandboxError]:
        return await posix.import_tree(self, tar, context)

    async def measured(self, context: SandboxContext) -> Ok[str] | Err[SandboxError]:
        """The manifest hash of /workspace through its export; an archive the sandbox
        answered that the reader refuses is unavailable."""
        hashed = await tree_hash(self, context)
        if isinstance(hashed, Ok):
            return hashed
        failed = hashed.error
        if isinstance(failed, SandboxError):
            return Err(failed)
        return Err(SandboxError("unavailable", failed.message))

    async def snapshot(
        self, operation_key: str, context: SandboxContext
    ) -> Ok[SnapshotData] | Err[SandboxError]:
        """Only a provider-owned whole-sandbox boundary makes a capture quiescent. The tree
        before and after it must also match, or a writer ran around it (not_quiescent). An
        A->B->A writer still passes here: core proves the image by a ledgered restore."""
        capture = self._driver.capture
        if capture is None or capture.quiescence == "unconfirmed":
            message = f"{self._provider} has no confirmed quiescent snapshot"
            return Err(SandboxError("unavailable", message))
        before = await self.measured(context)
        if isinstance(before, Err):
            return before
        taken = await call(self._driver, context, lambda: capture.take(self._id, operation_key))
        if isinstance(taken, Err):
            return taken
        after = await self.measured(context)
        hashed = before.value
        if isinstance(after, Ok) and after.value == hashed:
            return Ok(self._data(capture, taken.value, hashed))
        await call(self._driver, context, lambda: capture.delete(taken.value.ref))
        if isinstance(after, Err):
            return after
        return Err(SandboxError("not_quiescent", f"the tree of {self._id} changed around it"))

    async def close(self, context: SandboxContext) -> Ok[None] | Err[SandboxError]:
        killed = await call(self._driver, context, lambda: self._driver.kill(self._id))
        return released(killed.error) if isinstance(killed, Err) else Ok(None)

    def _data(self, capture: Capture, taken: Taken, hashed: str) -> SnapshotData:
        # The boundary covered the whole sandbox: every process in it was frozen, or ended.
        whole = [self._id]
        paused = capture.quiescence == "paused"
        return SnapshotData.model_validate(
            {
                "snapshot_id": taken.ref,
                "provider": self._provider,
                "sandbox_id": self._id,
                "capture_class": "filesystem",
                "expires_at": taken.expires_at,
                "manifest_hash": hashed,
                "quiesced": {
                    "frozen": whole if paused else [],
                    "stopped": [] if paused else whole,
                    "excluded": [],
                },
            }
        )
